src/jdk.jartool/share/classes/sun/security/tools/jarsigner/Main.java

```diff
@@ -1964,6 +1964,30 @@ void signJar(String jarName, String alias)
 
         Throwable failedCause = null;
         String failedMessage = null;
+        JarFile asJar = null;
+
+        try {
+            asJar = new JarFile(jarFile);
+        } catch (IOException ioe) {
+            // intentionally "eat" this, since we don't want to fail, if we
+            // cannot perform the multiple manifest check to output the warning
+        }
+
+        // only perform this check if the ZipFile is a JarFile
+        if(asJar != null && verbose != null) {
+            int manifestNum = JUZFA.getManifestNum(asJar);
+            try {
+                asJar.close();
+            } catch (IOException ioe) {
+                // intentionally "eat" this, since we don't want to fail
+            }
+            if(manifestNum > 1) {
+                System.out.println();
+                System.out.println(rb.getString("Warning."));
```

Comment:
Typically, we don’t print a warning immediately when a problem is detected. Instead, we log a flag and display the warnings later in the `displayMessagesAndResult` method. If a warning is printed too early, and there are a lot of additional lines after it (For example, with `-verbose` enabled and many entries), the warning may be overlooked.

Reply:
I see. I noted the `error` method but somehow I missed that warnings are collected and displayed via `displayMessagesAndResult`. I was planning to show the message as it happens, since that is what we do with the `error` method calls. This isn't exactly an "error," but I didn't see a `warn` (or similar) method.

Reply:
It uses displayMessagesAndResult() where it will keep all of the warnings at https://github.com/openjdk/jdk/blob/master/src/jdk.jartool/share/classes/sun/security/tools/jarsigner/Main.java#L1134 and emit them as needed later in this method.